### ISSUE TITLE
How to mock with

### ISSUE THREAD

**Issue by NickSeagull** [commercialhaskell/rio#211]
In other libraries, one would parameterize the monad in the handle type to then mock it in any monad, like identity.

But here we use `Has* env` classes in order to access them, the monad is always `RIO`. Could you provide some insights please? 🙏

**Comment by snoyberg** [comment]
Why do you need identity to mock?

**Comment by NickSeagull** [comment]
Identity was an example, let's suppose that I wanted to use `State` to see that everything was called in order, or perhaps using [`mockazo`](https://github.com/theam/mockazo) to mock automatically

**Comment by snoyberg** [comment]
I don't have experience with that package. Mocking can be handled effectively by using the `RIO` monad itself and providing mocked functions in the `Has` typeclass, such as a mocked log function that writes all messages to an `IORef`.

**Comment by NickSeagull** [comment]
Yes, that was my first approach, but what about handles that depend on each other? i.e.:

    data Server env = Server
      { run :: RIO env ()
      }
    
    class HasServer env where
      -- lens signature ...
    
    new :: HasHttp env => Server env
    new = Server
      { run = do
          Http{..} <- view httpL
          Database{..} <- view databaseL
          ...
      }

When initializing the environment with all of the services, I don't see a straight forward way of doing so, because I have to build the environment like:

    a <- Logger.new
    runRIO a $ do
    	b <- Server.new
      runRIO (a, b) $ do
        ...

Perhaps I'm missing something? The point of doing the services like that is to be able to do whitebox testing by mocking service functions out.

**Comment by snoyberg** [comment]
I'm afraid I don't understand the question, sorry.

**Comment by NickSeagull** [comment]
Essentially I'm trying to implement [what's on this post](http://felixmulder.com/writing/2019/10/05/Designing-testable-components.html). Which is the typical `ReaderT` application with a parameterized monad.

But I don't see how to implement it with RIO
